### DISCUSSION TITLE
<extension> in multiple <trk>'s in single gpx-file

### DISCUSSION THREAD

**Discussion by kosi2711** [osmandapp/OsmAnd#12983] [Q&A]
Hello alltogether,

when I have a collection of gpx files (ie from a hiking guide) I use shell scripts to concatenate them in one single file with waypoints to show an icon on the map and one \<trk\> with several \<trkseg\> to show the route on the map. That's working for several years now.

Osmand has great improvements! In newer releases it is possible to change color and symbol of an icon. So I modified my scripts:

- additional info in each \<wpt\>:

    \<extensions>
        \<osmand:color\>COLOR\</osmand:color\>
        \<osmand:icon\>special_trekking\</osmand:icon\>
        \<osmand:background\>square\</osmand:background\>
     \</extensions\>

- several \<trk\>

    Each single gpx file gets an own \<trk\> section.

This works (see Screenshots):

Depiction on map:

![Screenshot_20211012-165027](https://user-images.githubusercontent.com/82325892/136984682-44cfe4ee-3256-4d9a-b56c-9e76a8359a0c.png)

The single \<trk\>'s are recognized:

![Screenshot_20211012-165329](https://user-images.githubusercontent.com/82325892/136984762-3581ca8a-7030-40c1-b9ea-5a94e421e163.png)

Now my question: it would be (really!) great if I could change the color within each \<trk\>, so that not all displayed routes have the same color (green in the first screenshot). Therefore I inserted the corresponding statements to the \<trk\> section. 
 Here is what my scripts generate:

```
   ...
   </trkpt>
      <trkpt lat="48.463882" lon="8.411509">
        <ele>730.1</ele>
      </trkpt>
</trkseg>
</trk>
<!-- from ../../tracks-01/Nordschwarzwald/baiersbronner-himmelsweg-hoehen-tour.gpx -->
  <trk>
  <name>baiersbronner-himmelsweg-hoehen-tour.gpx</name>
  <extensions>
    <osmand:color>#323232</osmand:color>
    <osmand:icon>special_trekking</osmand:icon>
    <osmand:background>square</osmand:background>
  </extensions>
    <trkseg>
      <trkpt lat="48.532996" lon="8.289871">
        <ele>592.75646</ele>
      </trkpt>
      ...
```
(osmand:color changes in each \<trk\> section)

but that does not work - the color is always the same  (that one from the waypoints). 

There is no error generated ... Are the \<extension\> entries ignored?

Greets!

**Comment by vshcherb** [comment]
Good question why it doesn't work, cause ```color``` should be supported by each track or track segment.

**Comment by xdanielex** [comment]
This does work, and has for a while - so for anyone arriving here from a search, here is what the file has to look like.

The namespace has to be declared on the root gpx element. Without it the osmand: prefix is undefined, the extensions are silently skipped, and every track falls back to the same colour - which matches the symptom described above.

```xml
<gpx version="1.1" creator="whatever"
     xmlns="http://www.topografix.com/GPX/1/1"
     xmlns:osmand="https://osmand.net">
  <trk>
    <name>climb</name>
    <extensions>
      <osmand:color>#e02020</osmand:color>
      <osmand:width>12</osmand:width>
      <osmand:show_arrows>true</osmand:show_arrows>
      <osmand:coloring_type>solid</osmand:coloring_type>
    </extensions>
    <trkseg>
      <trkpt lat="46.140000" lon="11.085307"><ele>819.7</ele></trkpt>
      <trkpt lat="46.140283" lon="11.085556"><ele>820.9</ele></trkpt>
    </trkseg>
  </trk>
  <trk>
    <name>descent</name>
    <extensions>
      <osmand:color>#20a020</osmand:color>
      <osmand:width>12</osmand:width>
      <osmand:show_arrows>true</osmand:show_arrows>
      <osmand:coloring_type>solid</osmand:coloring_type>
    </extensions>
    <trkseg>
      <trkpt lat="46.141132" lon="11.086295"><ele>820.5</ele></trkpt>
      <trkpt lat="46.141415" lon="11.086544"><ele>818.2</ele></trkpt>
    </trkseg>
  </trk>
</gpx>
```

Two details that cost me time:

**coloring_type matters.** If it is left out, OsmAnd may fall back to a gradient mode and ignore the per-track colour. Setting it to `solid` makes the colour stick.

**Import as separate tracks, not "as one track".** With "Import as one track" the tracks are merged and the individual colours are lost. There was also a real bug here for a while: per-track colours broke in 4.9.10 and were fixed the following month (issue #21692). On a current version this only bites if you actually pick the "as one track" option.

Worth adding: with solid per-track colours the direction arrows stay visible, which they do not do in the built-in gradient colouring modes (issue #21082).

I ended up automating this to colour climbs and descents differently - the tool cuts a track into consecutive trk blocks and gives each one its own colour, following the direction of travel, so an uphill stays an uphill and reversing the route swaps the colours. MIT licensed, C++, no dependencies, in case it saves anyone the scripting:

https://github.com/xdanielex/gpx-slope-colors
